rs/replicated_state/src/replicated_state.rs

```diff
@@ -968,6 +968,67 @@ impl ReplicatedState {
         (expired_callback_count, errors)
     }
 
+    /// Enforces the best-effort message limit by repeatedly shedding the largest
+    /// best-effort messages of the canisters with the largest best-effort memory
+    /// usage until the memory usage drops below the limit.
+    ///
+    /// Time complexity: `O(n * log(n))`.
+    pub fn enforce_best_effort_message_limit(&mut self, limit: NumBytes) {
+        const ZERO: NumBytes = NumBytes::new(0);
+
+        // Check if we need to do anything at all before constructing a priority queue.
+        let mut memory_usage = self
+            .canister_states
+            .values()
+            .fold(NumBytes::new(0), |acc, canister| {
+                acc + canister.system_state.best_effort_message_memory_usage()
```

Comment:
Are we sure that `best_effort_message_memory_usage` is really `ZERO` if there are no messages? I remember in some places these estimates include sizes of some structs etc. which may also be non-zero if there is nothing in them.

Reply:
Right now `CanisterQueues::best_effort_message_memory_usage()` returns `MessagePool::best_effort_message_bytes` (see above), which is strictly a count of message bytes. It is technically possible that in the future we may include some overhead, although it's hard to see exactly what that may be or what it would be useful for.

Are you concerned about the way we build the priority queue below (where we include all canisters with non-zero best-effort memory usage)? I suppose I could code it a bit more defensively, so we (i) drop from the priority queue canisters that were unable to shed a message (whether that's because of a bug or because zero messages had non-zero memory usage); and (ii) we also stop iterating if the priority queue becomes empty. Then we have a theoretical possibility that we end up above the limit while there's nothing left to shed; but at least there's no way of getting into an infinite loop.

Reply:
Tweaked the implementation so it can deal with non-zero best-effort message memory usage for canisters that (for whatever reason) are unable to shed any best-effort message (whether it's because of a bug or because the memory usage includes some fixed overhead).

Also documented the `best_effort_message_memory_usage()` methods and `MessagePoolStats::best_effort_message_bytes` to explicitly say that the value is only non-zero if non-zero best-effort messages are pooled (and the other way around).